### PR TITLE
[Storage] `az storage copy`: Add positional argument `extra_options` to pass through options to `azcopy`

### PR DESCRIPTION
**Pull request by evelyn-ys**
**Description**

Feature asked by #20662 #20663 #20666 
This PR:
- Adds positional argument `extra_options` with `experimental` tag for flexibility
- Upgrades `azcopy` from `10.8.0` to `10.13.0`
- Removes `preview` tag for `az storage copy`

**Testing Guide**

`az storage copy -s source_file -d destination_container_url -- --block-size-mb=0.25 --check-length`

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by jiasli** [comment]
Unfortunately, there is not easy solution for argument pass-though. Apparently, you are choosing the 1st workaround from https://github.com/Azure/azure-cli/issues/18869.

**Comment by jiasli** [comment]
Unfortunately, there is no easy solution for argument pass-though. Apparently, you are choosing the 1st workaround from https://github.com/Azure/azure-cli/issues/18869. This makes passing quotes `"` `'` complicated as they need to be quoted in shell, such as

```sh
--extra-options "--some-other-parameter \"value with space\""
```

or even worse with single quotes both inside and outside

```sh
--extra-options '--some-other-parameter '"'"'value with space'"'"' something_else'
                 ^^^^^^^^^^^^^^^^^^^^^^^  ^  ^^^^^^^^^^^^^^^^  ^  ^^^^^^^^^^^^^^^  # string content

$ echo '--some-other-parameter '"'"'value with space'"'"' something_else'
--some-other-parameter 'value with space' something_else
```

**Comment by jiasli** [comment]
Things in PowerShell gets worse as now you need to consider escaping 3 times from

1. `az`
2. PowerShell
3. CMD

This becomes unmanageably complicated. 

See https://github.com/Azure/azure-cli/blob/dev/doc/quoting-issues-with-powershell.md

**Comment by jiasli** [comment]
Here is the script I use to debug argument passing in shells:

```py
# echo_args.py

import sys


def echo(args):
    print('[str]')
    for i, arg in enumerate(args):
        print(f"{i}: {arg!s}")

    print()

    print('[repr]')
    for i, arg in enumerate(args):
        print(f"{i}: {arg!r}")


echo(sys.argv)
```

You can invoke it by

```
> python .\echo_args.py --extra-options "--some-other-param \`"value with \\\`"quote\\\`"\`""
[str]
0: .\echo_args.py
1: --extra-options
2: --some-other-param "value with \"quote\""

[repr]
0: '.\\echo_args.py'
1: '--extra-options'
2: '--some-other-param "value with \\"quote\\""'
```

File: src/azure-cli/azure/cli/command_modules/storage/_params.py [hunk 1]
```diff
@@ -1076,6 +1076,8 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
         c.argument('cap_mbps', arg_group='Additional Flags', help="Caps the transfer rate, in megabits per second. "
                    "Moment-by-moment throughput might vary slightly from the cap. "
                    "If this option is set to zero, or it is omitted, the throughput isn't capped. ")
+        c.argument('extra_options', arg_group='Additional Flags',
+                   help="Other options which will be passed through to azcopy as it is")
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/storage/_params.py:1080]: Let's include some examples in the help message.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/storage/_params.py:1080]: If we can't decide which way from #18869 to go, consider mark it as experimental.

File: src/azure-cli/azure/cli/command_modules/storage/_params.py [hunk 2]
```diff
@@ -1076,6 +1076,8 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
         c.argument('cap_mbps', arg_group='Additional Flags', help="Caps the transfer rate, in megabits per second. "
                    "Moment-by-moment throughput might vary slightly from the cap. "
                    "If this option is set to zero, or it is omitted, the throughput isn't capped. ")
+        c.argument('extra_options', arg_group='Additional Flags',
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/storage/_params.py:1079]:
```suggestion
        c.argument('extra_options', arg_group='Additional Flags', is_experimental=True,
```

File: src/azure-cli/azure/cli/command_modules/storage/_params.py [hunk 3]
```diff
@@ -1076,6 +1076,9 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
         c.argument('cap_mbps', arg_group='Additional Flags', help="Caps the transfer rate, in megabits per second. "
                    "Moment-by-moment throughput might vary slightly from the cap. "
                    "If this option is set to zero, or it is omitted, the throughput isn't capped. ")
+        c.positional('extra_options', nargs='*', is_experimental=True, default=[],
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/storage/_params.py:1079]: I think `[]` is already the default:

```py
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('extra', nargs="*")

print(parser.parse_args(''.split()))

>>> Namespace(extra=[])
```

**Comment by jiasli** [comment]
Copying my comment from https://github.com/Azure/azure-cli/issues/20663#issuecomment-993243242

This topic was intensively discussed before in https://github.com/Azure/azure-cli/issues/18869.

Then I would propose the `--` solution:

```
az storage copy -s source_file -d container_url -- --bla="bla" --foo="foo" --bar="bar"
```

This is a common convention in Linux world:

- https://unix.stackexchange.com/questions/11376/what-does-double-dash-mean
- https://stackoverflow.com/questions/13321458/meaning-of-git-checkout-double-dashes

@chasewilson for awareness. If we go down this path, then other commands can do the same.

**Comment by chasewilson** [comment]
> Copying my comment from [#20663 (comment)](https://github.com/Azure/azure-cli/issues/20663#issuecomment-993243242)
> 
> This topic was intensively discussed before in #18869.
> 
> Then I would propose the `--` solution:
> 
> ```
> az storage copy -s source_file -d container_url -- --bla="bla" --foo="foo" --bar="bar"
> ```
> 
> This is a common convention in Linux world:
> 
> * https://unix.stackexchange.com/questions/11376/what-does-double-dash-mean
> * https://stackoverflow.com/questions/13321458/meaning-of-git-checkout-double-dashes
> 
> @chasewilson for awareness. If we go down this path, then other commands can do the same.

I think this makes a lot of sense all things considered. I'm glad this is experimental as I'm hesitent to make this the standard for similar scenarios. From a command line perspective, it's less than ideal to have a "native" command like this where you have to know another command line tools (azcopy) parameters to make the most of it. 

Is there are reason we don't support wrapping the new flags and parameters. Please forgive me if i'm catching up and I'm not saying we shouldn't implement the current solution, I just want to understand all avenues before committing. 

Another question, are we relying on the native bash parsing for the `--` mangagement? If so, how will we approach from Windows?

Thank you!

**Comment by jiasli** [comment]
> Another question, are we relying on the native bash parsing for the `--` mangagement? If so, how will we approach from Windows?

`--` interpretation is provided by Python's built-in lib `argparse`, so yes, it works on Windows too.

**Comment by evelyn-ys** [comment]
> Is there are reason we don't support wrapping the new flags and parameters.

There has been and will be more and more `azcopy` options in new versions. It's kind of duplicated work for CLI to add all the options and release new version. So we would like pass-through arguments. But if you think it's safer to wrap them all, I can also understand.

File: src/azure-cli/azure/cli/command_modules/storage/_params.py (hunk 3 above)
Comment (evelyn-ys) [reply src/azure-cli/azure/cli/command_modules/storage/_params.py:1079]: I didn't know and added manually. Thanks for notifying.

**Review by jiasli — APPROVED** [review]

**Review by calvinhzy — APPROVED** [review]

**Pull request merged**